### DISCUSSION TITLE
Path is created incorrectly.

### DISCUSSION THREAD

**Discussion by haefele** [cake-build/cake#3394] [Q&A]
Let's assume a project structure like the following:
- Root
  - MyProjectRoot
    - src
      - MyProject.sln
      - ...
    - tools
      - cake
        - **cake.exe**
        - ...
    - scripts
      - **build.cake**
      - **build.ps1**

My **build.ps1** script looks like:

```
cd ..\tools\cake
.\Cake.exe .\..\..\scripts\build.cake -verbosity=diagnostic
```

I think my intention is clear: I want to execute my **build.cake** script.

But it looks like he creates the full-path to the **build.cake** file incorrectly.
The output is: 

```
Creating script session...
Processing build.cake...
Could not find script "Root/scripts/build.cake".
```

It looks like he initially found my **build.cake** file, but later he creates the path incorrectly and obviously can't find it at the new path.
This is one folder too much upwards in the folder hierarchy.

So then I try to remove one **\..** from my path to **build.cake**:

```
cd ..\tools\cake
.\Cake.exe .\..\scripts\build.cake -verbosity=diagnostic
```

But sadly, this doesn't help either.
Because now, he fails with another error.
The output is: 

```
Could not find a part of the path "MyProjectRoot/tools/scripts".
```

Okay, I didn't really expect **that** to work now, since that path can't be correct, it's one folder too much downwards in the folder hierarchy.

So it looks like, I have no way to get into my "MyProjectRoot" folder, to point the cake.exe to the correct script file.

**Comment by patriksvensson** [comment]
Hmm, sounds strange. Will take a look at it later tonight.
Thank you for reporting this.

**Comment by patriksvensson** [comment]
@haefele I just tried this and it works as expected here. Would you mind posting your `build.ps1` script here? Would help reproduce the behavior.

**Comment by haefele** [comment]
@patriksvensson Well, they are exactly like the ones above.
I will try to reproduce it on my end now.

**Comment by haefele** [comment]
@patriksvensson This is strange, it still doesn't work on my end.
FYI, I have version 0.1.32.0, the latest build I can get of github.

Maybe I can upload my folders somewhere?

**Comment by patriksvensson** [comment]
@haefele I just used the Cake Bootstrapper to generate a script and edited that one to use the folder structure you have. 

Here is the script, try it out and see if it works on your end.

```
Param(
    [string]$Script = "build.cake",
    [string]$Target = "Default",
    [string]$Configuration = "Release",
    [string]$Verbosity = "Verbose"
)


$TOOLS_DIR = Join-Path $PSScriptRoot "../tools"
$NUGET_EXE = Join-Path $TOOLS_DIR "nuget.exe"
$CAKE_EXE = Join-Path $TOOLS_DIR "Cake/Cake.exe"
$PACKAGES_CONFIG = Join-Path $TOOLS_DIR "packages.config"

# Make sure tools folder exists
if ((Test-Path $PSScriptRoot) -and !(Test-Path $TOOLS_DIR)) {
    New-Item -path $TOOLS_DIR -name logfiles -itemtype directory
}

# Try find NuGet.exe in path if not exists
if (!(Test-Path $NUGET_EXE)) {
    "Trying to find nuget.exe in path"
    $NUGET_EXE_IN_PATH = &where.exe nuget.exe
    if ((Test-Path $NUGET_EXE_IN_PATH)) {
        "Found $($NUGET_EXE_IN_PATH)"
        $NUGET_EXE = $NUGET_EXE_IN_PATH 
    }
}

# Try download NuGet.exe if not exists
if (!(Test-Path $NUGET_EXE)) {
    Invoke-WebRequest -Uri http://nuget.org/nuget.exe -OutFile $NUGET_EXE
}

# Make sure NuGet exists where we expect it.
if (!(Test-Path $NUGET_EXE)) {
    Throw "Could not find NuGet.exe"
}

# Save nuget.exe path to environment to be available to child processed
$ENV:NUGET_EXE = $NUGET_EXE

# Restore tools from NuGet.
Push-Location
Set-Location $TOOLS_DIR

# Restore packages
if (Test-Path $PACKAGES_CONFIG)
{
    Invoke-Expression "$NUGET_EXE install -ExcludeVersion"
}
# Install just Cake if missing config
else
{
    Invoke-Expression "$NUGET_EXE install Cake -ExcludeVersion"
}
Pop-Location
if ($LASTEXITCODE -ne 0)
{
    exit $LASTEXITCODE
}

# Make sure that Cake has been installed.
if (!(Test-Path $CAKE_EXE)) {
    Throw "Could not find Cake.exe"
}

# Start Cake
Invoke-Expression "$CAKE_EXE `"$Script`" -target=`"$Target`" -configuration=`"$Configuration`" -verbosity=`"$Verbosity`""
exit $LASTEXITCODE
```

**Comment by haefele** [comment]
@patriksvensson Your script seems to work, but I can't tell why mine is not.
Shall I upload my folders and scripts somewhere, so you can easily test my case?

**Comment by patriksvensson** [comment]
Yes, that would be good. There must be a rational explanation for this :)

**Comment by haefele** [comment]
I hope you can work with a onedrive file :smiley:.
http://1drv.ms/1Ax3uNX

**Comment by patriksvensson** [comment]
Perfect. The problem seem to have something to do how Powershell handles working directories when launching a process like that. Maybe @devlead know something about this?

Anyway, you can get your script working correctly by running it like this:

``` powershell
.\..\tools\cake\Cake.exe build.cake -verbosity=diagnostic
```

**Comment by haefele** [comment]
This is just so strange, because I can get one folder upwards or downwards of the actual folder I want to get to. 
Anyways, that seems to do the trick. 

Thank you very much!

And especially thanks for the amazing fast help!

**Comment by patriksvensson** [comment]
Will investigate this further to make sure that this is not a problem with Cake.
Anyways, glad that you got it running!
